state-chain/runtime/src/chainflip.rs

```diff
@@ -341,16 +345,39 @@ impl TransactionBuilder<Solana, SolanaApi<SolEnvironment>> for SolanaTransaction
 	}
 
 	fn requires_signature_refresh(
-		_call: &SolanaApi<SolEnvironment>,
+		call: &SolanaApi<SolEnvironment>,
 		_payload: &<<Solana as Chain>::ChainCrypto as ChainCrypto>::Payload,
-		_maybe_current_onchain_key: Option<<SolanaCrypto as ChainCrypto>::AggKey>,
-	) -> bool {
-		// We use Durable Nonce mechanism to avoid the 150 blocks expiry period and so
-		// transactions won't expire. Then, the only reason to resign would be if the
-		// payload has been updated or the aggKey has been updated (key rotation).
-		// The payload won't be refreshed, as explained above, and the the broadcast
-		// barrier prevents a transaction from requiring to be resigned by a new aggKey.
-		false
+		maybe_current_onchain_key: Option<<SolanaCrypto as ChainCrypto>::AggKey>,
+	) -> RequiresSignatureRefresh<SolanaCrypto, SolanaApi<SolEnvironment>> {
+		// The only reason to resign would be if the aggKey has been updated on chain (key rotation)
+		// and this apicall is signed with the old key and is still pending. In this case, we need
+		// to modify the apicall, by replacing the aggkey with the new key, in the key_accounts in
+		// the tx's message to create a new valid threshold signing payload.
+		maybe_current_onchain_key.map_or(RequiresSignatureRefresh::False, |current_on_chain_key| {
+			match call.signer().and_then(|signer| {
+				if signer != current_on_chain_key {
+					Some(signer)
+				} else {
+					None
+				}
+			}) {
+				Some(signer) => {
+					let mut modified_call = (*call).clone();
+					let current_aggkey = SolanaThresholdSigner::keys(
+						SolanaThresholdSigner::current_key_epoch().unwrap(),
+					)
+					.unwrap();
```

Comment:
I think this is fine, but these two unwraps should be safe, right?

Reply:
Yes, they should be safe. Will add a comment explaining why